pkg/clustertree/cluster-manager/cluster_controller.go

```diff
@@ -0,0 +1,210 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	clusterlinkv1alpha1 "github.com/kosmos.io/kosmos/pkg/apis/kosmos/v1alpha1"
+	networkmanager "github.com/kosmos.io/kosmos/pkg/clusterlink/network-manager"
+	"github.com/kosmos.io/kosmos/pkg/clustertree/cluster-manager/controllers"
+	"github.com/kosmos.io/kosmos/pkg/clustertree/cluster-manager/utils"
+	"github.com/kosmos.io/kosmos/pkg/scheme"
+)
+
+const (
+	ControllerName = "cluster-controller"
+	//RequeueTime    = 5 * time.Second
+
+	ControllerFinalizerName      = "kosmos.io/cluster-manager"
+	MasterClusterAnnotationKey   = "kosmos.io/cluster-role"
+	MasterClusterAnnotationValue = "master"
+
+	DefaultClusterKubeQPS   = 40.0
+	DefalutClusterKubeBurst = 60
+)
+
+type ClusterController struct {
+	Master        client.Client
+	EventRecorder record.EventRecorder
+	Logger        logr.Logger
+
+	// clusterName: Manager
+	ControllerManagers     map[string]*manager.Manager
+	ManagerCancelFuncs     map[string]*context.CancelFunc
+	ControllerManagersLock sync.Mutex
+}
+
+func isMasterCluster(cluster *clusterlinkv1alpha1.Cluster) bool {
+	annotations := cluster.GetAnnotations()
+	if val, ok := annotations[MasterClusterAnnotationKey]; ok {
+		return val == MasterClusterAnnotationValue
+	}
+	return false
+}
+
+var predicatesFunc = predicate.Funcs{
+	CreateFunc: func(createEvent event.CreateEvent) bool {
+		obj := createEvent.Object.(*clusterlinkv1alpha1.Cluster)
+		return !isMasterCluster(obj)
+	},
+	UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+		obj := updateEvent.ObjectNew.(*clusterlinkv1alpha1.Cluster)
+		old := updateEvent.ObjectOld.(*clusterlinkv1alpha1.Cluster)
+
+		if isMasterCluster(obj) {
+			return false
+		}
+
+		// For now, only kubeconfig & DeletionTimestamp changes are concerned
+		if !bytes.Equal(old.Spec.Kubeconfig, obj.Spec.Kubeconfig) {
+			return true
+		}
+
+		if old.DeletionTimestamp != obj.DeletionTimestamp {
+			return true
+		}
+
+		return false
+	},
+	DeleteFunc: func(deleteEvent event.DeleteEvent) bool {
+		obj := deleteEvent.Object.(*clusterlinkv1alpha1.Cluster)
+		return !isMasterCluster(obj)
+	},
+	GenericFunc: func(genericEvent event.GenericEvent) bool {
+		return false
+	},
+}
+
+func (c *ClusterController) SetupWithManager(mgr manager.Manager) error {
+	c.ManagerCancelFuncs = make(map[string]*context.CancelFunc)
+	c.ControllerManagers = make(map[string]*manager.Manager)
+	c.Logger = mgr.GetLogger()
+
+	return controllerruntime.NewControllerManagedBy(mgr).
+		Named(ControllerName).
+		WithOptions(controller.Options{}).
+		For(&clusterlinkv1alpha1.Cluster{}, builder.WithPredicates(predicatesFunc)).
+		Complete(c)
+}
+
+func (c *ClusterController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	klog.V(4).Infof("============ %s starts to reconcile %s ============", ControllerName, request.Name)
+	defer func() {
+		klog.V(4).Infof("============ %s has been reconciled =============", request.Name)
```

Comment:
I think it is not suitable to place a defer function for logging that will make a mis understanding

Reply:
This is not a big problem, I will optimize it next time